### ISSUE TITLE
Streaming compression format

### ISSUE THREAD

**Issue by Yoric** [binast/binjs-ref#140]
This is a braindump of the streaming compression format.

**FIXME** Specify scope-based Debruijn-style indexing.

# Collapsing the tree: TreeRePair

The role of TreeRePair is to collapse the tree in a smaller tree, at the cost of a larger variety of nodes.

## Compression

The general idea of [TreeRePair](http://www.eti.uni-siegen.de/ti/veroeffentlichungen/12-repair.pdf) is the following:

- if it is common to have a node labelled `A` with a child labelled `B` in position `n`, then
   - create a new label `N`
   - insert in the dictionary `N = (A, n, B)` – by definition, `N.rank = A.rank + B.rank - 1`;
   - for each instance `a` of `A` with a child `b` labelled `B` in position `n` with `N`;
      - replace the label of `a` with `N`;
      - replace the child `b` with all its own children;
- proceed with the next-most common relationship
- continue as long as this decreases the size of the tree (possibly a command-line parameter)

## Decompression

Whenever we encounter a node `c` with label `C`:

- if `C` is part of the grammar, proceed;
- if `C` is not part of the grammar,
  - it must be in the dictionary, defined as `C = (A, n, B)`;
  - replace the label of `c` with `A`;
  - read `n - 1` nodes as the `n - 1` first children of `c`;
  - create a new node `b`, initially empty;
  - read `B.rank` nodes as the children of `b`;
  - `b` becomes the `n`th child of `c`;
  - read `A.rank - n` children as the remaining children of `c`.

# Efficient binary representation: multistream

The role of multistream is to represent each node of the AST in a manner that can efficiently be represented by brotli/gzip. We segregate across stream by type of data, in order to minimize entropy.

We have the following streams:

- tree structure;
- tree lengths;
- tree references;
- node definition type;
- grammar node definition length;
- grammar node definition content;
- generated node index;
- generated node parent/child;
- string definitions contents;
- string definition lengths;
- string references;
- identifier definitions contents;
- identifier definition lengths;
- identifier references;
- booleans;
- list lengths;
- numbers;

**Note** In the future, we may want to support additional (external) streams, e.g. sourcemaps, comments.

### Encoding the (extended) grammar

Whenever we first encounter a node type:
- if this is a grammar node:
  - write `0` to the stream *node definition types*;
  - write the number of bytes in the name to the stream *grammar node definition length*;
  - write the UTF-8 encoded bytes of the name to the stream *grammar node definition contents*;
- otherwise it is a TreeRePair digram `(A, n, B)`;
  - write `1` to the stream *node definition types*;
  - write `n` to the stream *generated node index*;
  - write a representation of `A` to the stream *generated node parent/child* (note: since `A` is a node type, this might be the first time we encounter it, so this may recurse);
  - write a representation of `B` to the stream *generated node parent/child*  (note: since `B` is a node type, this might be the first time we encounter it, so this may recurse);

TBD: Specify the "representation of ...".

### Encoding a node

- if this is the first time we encounter the node type:
  - we need to add it to the dictionary, see **Encoding the (extended) grammar
- write the index of the node to stream *tree structure*;
- if the node is skippable:
  - add the node to the FIFO *trees to encode*;
  - write the index of the skippable to stream *tree references*;
- else:
  - encode children.

TBD: Specify the index of the node. Last time I checked, the best encoding was using Parent Prediction.

### Encoding a string enum value

String enum values are handled as nodes, e.g. a node `+`.

TBD: Specify how we handle collisions. Should we rather write `BinaryOperation::+`?

### Encoding a raw string (not string enums, not identifiers)

- if this is the first time we encounter a string:
  - if this is the `null` string:
    - substitute the invalid string `0xFF;0x00` (2 bytes);
  - write the number of bytes of the string to stream *string definition lengths*;
  - write the utf-8 encoded bytes of the string to stream *string definition contents*;
- write the index of the string to stream *string references*

TBD: Specify the index of the string.

### Encoding an identifier

- if this is the first time we encounter an identifier:
  - write the number of bytes of the string to stream *identifier definition lengths*;
  - write the utf-8 encoded bytes of the string to stream *identifier definition contents*;
- write the index of the string to stream *identifier references*

TBD: Specify the index of the identifier. Last time I checked, setting up identifiers per scope was the best encoding.

### Encoding a boolean

- if the boolean is `true`
  - write `1` to the stream *booleans*
- else if the boolean is `false`
  - write `0` to the stream *booleans*
- else if the boolean is `null`
  - write `2` to the stream *booleans*

### Encoding a list

- if the list is `null`
  - write invalid varnum `0x01;0x00` to the stream *list lengths*;
- else:
  - write the u32 varnum to the stream *list lengths*;
  - encode the child nodes

### Encoding a number

- if the number is `null`:
  - write invalid varnum `0x01;0x01;0x00` to the stream *numbers*;
- else if the number is a `i32`:
  - cast number to `u32`;
  - write number as a varnum to the stream *numbers*;
- else (number needs a `f64`):
  - write invalid varnum `0x01;0x00` to the stream *numbers*;
  - write 8 bytes of the representation of number to the stream *numbers*.

### Once a tree is done encoding

- write the number of bytes to stream *tree lengths*;
- forget current scope;
- pop (scope, tree) from FIFO trees to encode;

## Decompression

TBD

# Interleaving

## Compression

To allow streaming parsing/compilation, we need all the streams to be delivered at the same time. For this purpose, we cut the data into packets every so often. Given a target size N, where N is probably a command-line parameter, whenever we realize that we have written N bytes, we flush all streams and put them all in a packet.

TBD

## Decompression

TBD
